ff/src/fields/arithmetic.rs

```diff
@@ -31,25 +64,42 @@ macro_rules! impl_field_mul_assign {
                         return;
                     }
                 }
-                let mut r = [0u64; $limbs];
-                let mut carry1 = 0u64;
-                let mut carry2 = 0u64;
-
-                for i in 0..$limbs {
-                    r[0] = fa::mac(r[0], (self.0).0[0], (other.0).0[i], &mut carry1);
-                    let k = r[0].wrapping_mul(P::INV);
-                    fa::mac_discard(r[0], k, P::MODULUS.0[0], &mut carry2);
-                    for j in 1..$limbs {
-                        r[j] = mac_with_carry!(r[j], (self.0).0[j], (other.0).0[i], &mut carry1);
-                        r[j - 1] = mac_with_carry!(r[j], k, P::MODULUS.0[j], &mut carry2);
+                self.mul_assign_nocarry(other);
```

Comment:
I believe the `asm` and the `mul_assign_nocarry` algorithms are the same, so we can just move the `asm` into `mul_assign_nocarry` directly.